src/runtime/cuda/gpu_cache.cu

```diff
@@ -0,0 +1,166 @@
+/*!
+ *  Copyright (c) 2022 by Contributors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ * \file gpu_cache.cu
+ * \brief Implementation of wrapper HugeCTR gpu_cache routines.
+ */
+
+#ifndef DGL_ARRAY_CUDA_GPU_CACHE_H_
+#define DGL_ARRAY_CUDA_GPU_CACHE_H_
+
+#include <dgl/packed_func_ext.h>
+#include <dgl/runtime/registry.h>
+#include <dgl/runtime/object.h>
+#include <dgl/array.h>
+#include <dgl/aten/array_ops.h>
+#include <dgl/runtime/container.h>
+#include <dgl/runtime/device_api.h>
+#include <cuda_runtime.h>
+#include "../../runtime/cuda/cuda_common.h"
+
+#include <nv_gpu_cache.hpp>
+
+namespace dgl {
+namespace array {
+namespace cuda {
+
+template <typename key_t>
+class GpuCache : public runtime::Object {
+constexpr static int set_associativity = 2;
+constexpr static int WARP_SIZE = 32;
+constexpr static int bucket_size = WARP_SIZE * set_associativity;
+
+ public:
+    static constexpr const char* _type_key =
+                    sizeof(key_t) == 4 ? "cuda.GpuCache32" : "cuda.GpuCache64";
```

Comment:
IIRC, C++/CUDA source code in DGL usually uses the indentation of 2 spaces, and another 4 spaces for line-break in a continuous sentence.